proto/kv-connect.md

```diff
@@ -0,0 +1,448 @@
+requests a `x-denokv-version` header with a value of the protocol version that
+the client is using.
+
+The server MUST NOT respond with 3xx class redirect responses to data path
+requests.
+
+There are two types of requests that the client may send to the server:
+
+- A _Snapshot Read Request_ which is used to read some values from the database.
+- An _Atomic Write Request_ which is used to write some values to the database.
+
+#### Snapshot Read Request
+
+A _Snapshot Read Request_ is used to read some values from the database.
+
+<!-- A _Snapshot Read Request_ can either be "single shot" or "watching". A "single
+shot" _Snapshot Read Request_ is used to read a given set of values from the
+database once. A "watching" _Snapshot Read Request_ is used to read a given set
+of values from the database once and then receive a stream of  -->
+
+A _Snapshot Read Request_ can either be strongly consistent or eventually
+consistent. A strongly consistent _Snapshot Read Request_ MUST be sent to an
+endpoint that has a `consistency` property that is equal to `strong`. An
+eventually consistent _Snapshot Read Request_ MUST be sent to an endpoint that
+has a `consistency` property that is either equal to `strong` or `eventual`.
+
+To determine which endpoint to send a _Snapshot Read Request_ to, the client
+MUST first narrow down the list of endpoints to only include endpoints that have
+a valid consistency property for the type of _Snapshot Read Request_ that the
+client wants to perform. From this list, the client MAY make requests to any
+endpoint, but the client SHOULD prefer endpoints that have lower latency.
+
+The client sends a _Snapshot Read Request_ to the server by sending a POST
+request to the `<endpointUrl>/snapshot_read` endpoint, where `<endpointUrl>` is
+the resolved URL of the selected endpoint. The request body MUST be a Protobuf
+message in the format `com.deno.kv.datapath.SnapshotRead` (defined in the
+`schema/datapath.proto` file in this repository).
+
+<!-- If the client wants to receive a stream of updates from the server, rather than
+just a single response, the client MUST include a `x-denokv-stream` header with
+the value `true`. -->
+
+Upon receiving the POST request, the server MUST verify that the request body
+adheres to the Protobuf schema for the `com.deno.kv.datapath.SnapshotRead`
+message. If the request body is invalid, the server MUST respond with a 4xx
+response and a plain text body containing a human readable error message.
+
+The server MAY return a 5xx class HTTP status response at any time if the server
+is unable to process the request right now due to a temporary condition. The
+server SHOULD include a plain text body with a human readable error message in
+this case.
+
+The server MAY perform quota checks on the request. If the request is rejected
+due to quota limits, the server MUST respond with a 4xx class response and a
+plain text body containing a human readable error message.
```

Comment:
How about rate limiting?

Reply:
I have added a note to the paragraph above about using 5xx responses for rate limiting, as those are retried with backoff.